clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -435,6 +436,17 @@ public int compare(ProducerBatch o1, ProducerBatch o2) {
         inflightBatchesBySequence.get(batch.topicPartition).offer(batch);
     }
 
+    synchronized int firstInFlightSequence(TopicPartition topicPartition) {
```

Comment:
I know we don't do this consistently, but using `null` instead of a magic value like `-1` is a little better. Both in readability and catching errors. `Optional<Integer>` would be even better, but that requires Java 8. For example, we have some really hard to figure out bugs where we end up storing the high watermark as `-1`. This should have been caught earlier, but since it's simply an `Int`, it's easy to miss validation until it's too late.

Reply:
Also, we should always document such cases so that people can see it when reading the signature.

Reply:
I think null may be better in general, but in the `TransactionManager` class, we return simple types in the other similar methods. That actually has a benefit in these cases because the sentinel values are actually useful and don't have to be checked for explicitly. 

So I think it's best to be consistent within the class at least. I will update the javadoc.

Reply:
But you are checking it explicitly in the calling code. If the rest of the class is using this pattern, it's OK to keep it like it is for consistency (for now at least).